### ISSUE TITLE
[RFC] Refactor Python interface.

### ISSUE THREAD

**Issue by trivialfis** [dmlc/xgboost#5152]
# Scikit-Learn

Previously I use `xgboost.train` exclusively so many issues in Scikit-Learn interface are ignored.  With more and more users enjoying the convenience of SKL's ecosystem, it maybe better for us to put more focus on it.  Here are a few items in mind for future development:

- [x] Use callback to implement cross validation and early stopping, instead of accepting batches of parameters for validation data.
- [ ] Use Python `weak_ref` to implement a cache in Python side, this avoids calling constructor of `DMatrix`.
- [ ] Implement `staged_predict`.  With above item in place, we can use our native `DMatrix` cache to accelerate prediction on seen data.
- [x] Remove duplicated parameters.  mglowacki100 in https://github.com/dmlc/xgboost/pull/5130 showed one way to eliminate many duplicated parameters, although there are some questions still need to be answered, but I would very much like to focusing on merging it.
- [x] Model serialization.  With JSON now mainlined, we can utilize it for storing Scikit-Learn specific attributes.  See #4639 .
- [x] Unify `predict_prob` for `binary:logistic` and `multi:softprob`.
- [ ] Better support for custom objective, with support for weights and groups.

Expertise for SKL are welcomed, including suggestions for how to integrate XGBoost with SKL more smoothly, what's the better way to use various mixins etc.

# Callback

The implementation of callback functions is very "R style", I propose modelling our callback functions after [Keras](https://keras.io/callbacks/), this way we can remove all parameters related to cross validation, early stopping and use callback to implement these features.  Also it will be familiar to many Python data scientists.

At the same time, we can improve the call backs to handle both scikit-learn, native booster, dask in a unified way.

# Early stopping and training monitoring

XGBoost has its own set of functions handling early stopping that depends on Scikit-Learn.  I propose we remove them, and let users directly call Scikit-Learn interface.  It's way more popular and cleaner.

**Comment by RAMitchell** [comment]
Recommend avoiding extra state in the python layer if possible. DMatrix cache is a pain, I would want to see a strong use case for this before implementing.

**Comment by trivialfis** [comment]
@RAMitchell Sorry for the ambiguity here.  The cache is just used for avoiding reconstructing `DMatrix` in Python layer, not storing any internal state, so there is nothing needs to be synced.

**Comment by hcho3** [comment]
See https://discuss.xgboost.ai/t/ram-memory-keeps-increasing-with-iterative-training-of-xgb-models/1331/3 for an example of performance discrepancy between `xgboost.train()` and `XGBClassifier.fit()`.

**Comment by trivialfis** [comment]
Links:
https://scikit-learn.org/stable/developers/develop.html#rolling-your-own-estimator
https://scikit-learn.org/stable/modules/generated/sklearn.utils.estimator_checks.check_estimator.html

**Comment by trivialfis** [comment]
Closing as this is mostly done, evaluation metric has its own github issue.
